hydra-node/src/Hydra/Chain/CardanoClient.hs

```diff
@@ -317,12 +318,23 @@ queryStakePools networkId socket queryPoint =
    in runQuery networkId socket queryPoint query >>= throwOnEraMismatch
 
 -- | Throws at least 'QueryException' if query fails.
+-- We don't expect the query to fail but this can easily happen in case of a
+-- rollback. So what we do is retry 5 times and then give up.
 runQuery :: NetworkId -> SocketPath -> QueryPoint -> QueryInMode CardanoMode a -> IO a
 runQuery networkId socket point query =
-  queryNodeLocalState (localNodeConnectInfo networkId socket) maybePoint query >>= \case
-    Left err -> throwIO $ QueryAcquireException err
-    Right result -> pure result
+  let tracer = traceWith $ showTracing stdoutTracer
+   in go tracer 5
  where
+  go tracer n =
+    queryNodeLocalState (localNodeConnectInfo networkId socket) maybePoint query >>= \case
+      Left err ->
+        if n == (0 :: Int)
+          then throwIO $ QueryAcquireException err
+          else do
+            tracer ("Retrying query... " ++ show n <> " Query: " <> show query)
```

Comment:
This will make our logs non-JSON and hence it can only be temporary / blocks merging of this IMO.